File: hw3/mapping.py
import sys
import string

def readFile(file):
	mapping = dict()

	with open(file, 'r', encoding='big5-hkscs') as text:
		for line in text:
			first = line.split(' ')[0]
			wordList = line.split(' ')[1].split('/')
			if first not in mapping:
				mapping[first] = set(first)

			for i in range(len(wordList)):
				ZhuYin = wordList[i][0]
				if ZhuYin not in mapping:
					mapping[ZhuYin] = set(ZhuYin)
				else:
					mapping[ZhuYin].add(first)
	return mapping

def outputFile(fileName, mapping):
	file = open(fileName, 'w', encoding='big5-hkscs', newline=None)
	for c in mapping:
		file.write(c + ' ')
		for v in mapping[c]:
			file.write(v + ' ')		
		file.write("\n")
	file.close()

def main():
	mapping = readFile(sys.argv[1])
	outputFile(sys.argv[2], mapping)

if __name__ == "__main__":
	main()

File: hw3/mapping.cpp
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <vector>
using namespace std;

class DictionaryManager{
public:
    DictionaryManager(){}

    size_t getKey(const string& str, string& token, size_t start){
        token = str.substr(start, 2);
        start = str.find_first_of('/', start + 1); 
        return start;
    }
    
    void updateDict(string& key, string& word){
        if (key2Index.find(key) == key2Index.end())
        {
            vector<string> mapping{key};
            dictionary.push_back(mapping);
            pair<string, int> ZhuYin(key, index++);
            key2Index.insert(ZhuYin);
        }
        dictionary[key2Index.find(key)->second].push_back(word);
    }

    void processing(const string& str){
        string key, key_old;
        string word = str.substr(0, 2);
        size_t start = 3;

        int i = 0
        while(start != string::npos){
            start = getKey(str, key, start);
            if (key_old != key) { // handling '破音字'
                cout << key << "  " <<endl;
                updateDict(key, word);
                if
            }
        }
        cout <<endl;
        single_word.push_back(word);
    }

    void outputFile(const string& filename){
        ofstream ofs(filename.c_str(), ios::out);
        for(int i = 0; i < dictionary.size(); ++i)
        {
            ofs << dictionary[i][0]<<"   ";
            for(int j = 1; j < dictionary[i].size(); ++j){
                ofs<<' ' << dictionary[i][j];
            }
            ofs<<endl;
            
        }
        
        for (int i = 0; i < single_word.size(); ++i)
            ofs << single_word[i] << "    " << single_word[i] << endl;
    }

private:
    int index;
    unordered_map<string, int> key2Index;
    vector<vector<string> > dictionary;
    vector<string> single_word;
};

int main(int argc, char** argv){
    ifstream ifs;
    string line, key, word;
    string fileName = argv[1];
    DictionaryManager* mgr = new DictionaryManager;
    ifs.open(fileName.c_str(), ios::in);
    while(getline(ifs, line)){
        mgr->processing(line);
    }
    mgr->outputFile(argv[2]);
}